_posts/2016-05-25-coordination-in-akka.md

```diff
@@ -0,0 +1,147 @@
+
+> Using worker threads minimizes the overhead due to thread creation. Thread objects use a significant amount of memory, and in a large-scale application, allocating and deallocating many thread objects creates a significant memory management overhead.
+
+[Akka Actors](http://doc.akka.io/docs/akka/current/scala/actors.html) are extremely lightweight. We could afford creating millions of them if we knew that they would be disposed quickly. However that's not the case with our use case:
+
+<img src="/assets/img/custom/blog/fsm.png" alt="FSM architecture" title="FSM architecture" class="img img-center img-responsive style-screengrab">
+
+As you can see Finite State Machine (FSM) actors will wait until the Items service finishes deleting the items. It's important to note that the actor is waiting, but not blocking. Actors are attached to [Dispatchers](http://doc.akka.io/docs/akka/2.4.6/scala/dispatchers.html) that have a Thread Pool inside. Those thread pools have a really limited number of threads; blocking in one of the actors would mean running out of threads pretty quickly. Therefore our intent here is bounding the number of actors in our system, not threads. Maybe is an overkill for this particular example, but the point of this series is mainly educational. Let's see a diagram of our service with the proper level of abstraction.
+
+<img src="/assets/img/custom/blog/coordination-diagram.png" alt="Coordination" title="Coordination" class="img img-center img-responsive style-screengrab">
+
+## Akka HTTP
+
+[Akka HTTP](http://doc.akka.io/docs/akka/2.4.6/scala/http/introduction.html) is a library based on [Spray](http://doc.akka.io/docs/akka/2.4.6/scala/http/introduction.html) to create HTTP integration layers. The routing DSL that offers is extremely powerful and easy to read as you can see in our code:
+
+```scala
+  val deleteItem = get {
+    path("item" / JavaUUID) { itemId =>
+      onComplete(deleteItem(itemId)) {
+        case Success(Result(Right(_))) =>
+          complete(StatusCodes.OK)
+        case Success(Result(Left(Timeout(errorMessage)))) =>
+           complete(StatusCodes.GatewayTimeout)
+        // more cases   
+      }
+    }
+  }
+
+  private def deleteItem(itemId: UUID)= itemCoordinator ask ItemReported(itemId)
+```
+
+In this snippet I'd like to stress out **ask vs tell pattern**. The [tell pattern](http://doc.akka.io/docs/akka/2.4.6/scala/actors.html#Tell__Fire-forget) involves firing a message and forgetting about the response. With this approach information flows in one direction. However for our scenario we're interested in the [ask pattern](http://doc.akka.io/docs/akka/2.4.6/scala/actors.html#Ask__Send-And-Receive-Future). When we send a message to an actor through ask pattern a future is generated with an eventual response bounded by a timeout. As the documentation says:
```

Comment:
Do you want to just mention the distinction between these two patterns and the "Tell don't ask" principle? - see if you feel the need for it, but they are quite close as per the name goes, just to clarify meaning.